packages/core/src/actors.ts

```diff
@@ -0,0 +1,516 @@
+    transition: (_, event, { self, name }) => {
+      const { _parent: parent } = self;
+
+      if (event.type === startSignalType) {
+        const sender: Sender<TEvent> = (e) => {
+          if (canceled) {
+            return;
+          }
+
+          parent?.send(toSCXMLEvent(e, { origin: self }));
+        };
+
+        const receiver: Receiver<TEvent> = (newListener) => {
+          receivers.add(newListener);
+        };
+
+        dispose = invokeCallback(sender, receiver);
+
+        if (isPromiseLike(dispose)) {
+          dispose.then(
+            (resolved) => {
+              self._parent?.send(
+                toSCXMLEvent(doneInvoke(name, resolved), {
+                  origin: self
+                })
+              );
+
+              canceled = true;
+            },
+            (errorData) => {
+              const errorEvent = error(name, errorData);
+
+              self._parent?.send(
+                toSCXMLEvent(errorEvent, {
+                  origin: self
+                })
+              );
+
+              canceled = true;
+            }
+          );
+        }
+        return undefined;
+      } else if (event.type === stopSignalType) {
+        canceled = true;
+
+        if (isFunction(dispose)) {
+          dispose();
+        }
+        return undefined;
+      }
+
+      if (isSignal(event)) {
+        // TODO: unrecognized signal
+        return undefined;
+      }
+
+      const plainEvent = isSCXMLEvent(event) ? event.data : event;
+      receivers.forEach((receiver) => receiver(plainEvent));
+
+      return undefined;
+    },
+    initialState: undefined
+  };
+
+  return behavior;
+}
+
+export function fromPromise<T>(
+  lazyPromise: Lazy<PromiseLike<T>>
+): Behavior<{ type: string }, T | undefined> {
+  let canceled = false;
+  const resolveEventType = Symbol('resolve');
+  const rejectEventType = Symbol('reject');
+
+  // TODO: add event types
+  const behavior: Behavior<any, T | undefined> = {
+    transition: (state, event, { self, name }) => {
+      if (canceled) {
+        return state;
+      }
+
+      switch (event.type) {
+        case startSignalType:
+          const resolvedPromise = Promise.resolve(lazyPromise());
+
+          resolvedPromise.then(
+            (response) => {
+              self.send({ type: resolveEventType, data: response });
+            },
+            (errorData) => {
+              self.send({ type: rejectEventType, data: errorData });
+            }
+          );
+          return undefined;
+        case resolveEventType:
+          self._parent?.send(
+            toSCXMLEvent(doneInvoke(name, event.data) as any, {
+              origin: self
+            })
+          );
+          return event.data;
+        case rejectEventType:
+          const errorEvent = error(name, event.data);
+
+          self._parent?.send(
+            toSCXMLEvent(errorEvent, {
+              origin: self
+            })
+          );
+          return event.data;
```

Comment:
Curious on your thoughts here: since we want to represent an accurate promise snapshot, what do you think about representing promise snapshots as an "Either type" like other languages have?

Reply:
So based on T we'd type the actual snapshot as `{ ok: true, value: T } | { ok: false, error: unknown }` (or smth like that)?